### DISCUSSION TITLE
Strange behaviour from the `EVP_PKEY_decrypt` function

### DISCUSSION THREAD

**Discussion by Darnethal0z** [openssl/openssl#22838] [Q&A] [answered]
I recently made a script that encrypts and decrypts data with RSA with the EVP API.

It consists of 2 major functions : encryption and decryption (key pair are generated each run).

The encryption part works fine, but it is when i use the decryption part that the `EVP_PKEY_decrypt` function have a strange behaviour  : The parameter *outlen* is set to `0` after the function call.

According to the function manual page : 

> [...] The data to be decrypted is specified using the in and inlen parameters. If out is NULL then the minimum required size of the output buffer is written to the *outlen parameter.

Here is the decryption function that i'm using : 

```
int RSADecrypt(
    EVP_PKEY * pkey, 
    char * ciphertext, 
    char ** buffer, 
    unsigned int modulus){

    EVP_PKEY_CTX * ctx;
    char * decoded_buffer;
    size_t output_length;

    if (!(ctx = EVP_PKEY_CTX_new(pkey, NULL)))
        return ERROR_CTX_CREATION;
    if (!EVP_PKEY_encrypt_init(ctx))
        return ERROR_CTX_INIT;

    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_padding_mode", "oaep");
    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_oaep_md", "sha256");
    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_mgf1_md", "sha256");

    decoded_buffer = OPENSSL_malloc(modulus + 1);

    if(!decoded_buffer){
        EVP_PKEY_CTX_free(ctx);

        return -1;
    }

    // Base64 decoding
    if (EVP_DecodeBlock((unsigned char *) decoded_buffer, ciphertext, modulus) <= 0){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return ERROR_BASE64_ENCODE;
    }

    // Get the final output length first
    if (!EVP_PKEY_decrypt(ctx, NULL, &output_length, decoded_buffer, modulus)){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return ERROR_CTX_DECRYPTION;
    }

    printf("%i\n", output_length); // Displays 0

    *buffer = OPENSSL_malloc(output_length + 1);

    if (*buffer == NULL){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);
        
        return -1;
    }

    if (EVP_PKEY_decrypt(ctx, *buffer, &output_length, decoded_buffer, modulus) <= 0){
        OPENSSL_free(*buffer);
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return ERROR_CTX_DECRYPTION;
    }

    *buffer[output_length] = '\0';

    OPENSSL_free(decoded_buffer);
    EVP_PKEY_CTX_free(ctx);

    return 0;
}

```

The main function : 

```
int main(void){
        EVP_PKEY * pkey = EVP_RSA_gen(2048);

	char string[] = "HelloWorld";
	char * encrypted_buff;
	char * decrypted_buff;

        // ... The encryption part, working fine ...
        // The encrypted, base64 encoded string is in encrypted_buff

	// This is the decryption part

        int retval = RSADecrypt(pkey, encrypted_buff, &decrypted_buff, 256);
 
       if (retval != 0){
            free(encrypted_buff);
            free(decrypted_buff);
            printf("retval : %i\n", retval);
            return -1;
       }

	free(encrypted_buff);
        free(decrypted_buff);

	return 0;
}
```

I am certainly doing something wrong here, but i'll be looking forward for fixes.

Thanks in advance for your answers.

**Comment by nhorman** [comment, marked as answer]
I expect that the problem is that you initialized the context for an encrypt operation rather than a decrypt operation.  Doing so sets the ctx->operation value to EVP_PKEY_OP_ENCRYPT.  When you then call EVP_PKEY_decrypt, it compares the set operation to EVP_PKEY_OP_DECRYPT, and, when it doesn't match, returns -1.  

The documentation for EVP_PKEY_decrypt:
https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_decrypt.html
Indicates that a failure causes the function to return _either_ 0 or a negative value, but you only check for zero in the code above.  As such you skip your error condition, and continue on as if things had worked properly.

I imagine that if you change the call to EVP_PKEY_encrypt_init to be EVP_PKEY_decrypt_init, things will work better.

Hope that helps!

**Comment by Darnethal0z** [comment]
I successfully managed to fix my errors with a different approach.

I will not post the final function since it is now irrelevant to the issue. But if someone encounter the same problems as me, i'm posting here the fixes i brought to the `RSADecrypt()` function from the first message: 

```
int RSADecrypt(
    EVP_PKEY * pkey, 
    char * ciphertext, 
    char ** buffer, 
    unsigned int * out_len,
    unsigned int modulus){

    EVP_PKEY_CTX * ctx;
    char * decoded_buffer;
    size_t output_length;

    if (!(ctx = EVP_PKEY_CTX_new(pkey, NULL)))
        return -1;
    if (!EVP_PKEY_decrypt_init(ctx))
        return -2;

    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_padding_mode", "oaep");
    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_oaep_md", "sha256");
    EVP_PKEY_CTX_ctrl_str(ctx, "rsa_mgf1_md", "sha256");

    decoded_buffer = OPENSSL_malloc(modulus + 1);

    if(!decoded_buffer){
        EVP_PKEY_CTX_free(ctx);

        return -3;
    }

    // Base64 decoding
    if (EVP_DecodeBlock((unsigned char *) decoded_buffer, ciphertext, modulus) <= 0){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return -4;
    }

    // Get the final output length first
    if (!EVP_PKEY_decrypt(ctx, NULL, &output_length, decoded_buffer, modulus)){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return -5;
    }

    *buffer = OPENSSL_malloc(output_length + 1);

    if (*buffer == NULL){
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);
        
        return -3;
    }

    if (EVP_PKEY_decrypt(ctx, *buffer, &output_length, decoded_buffer, modulus) <= 0){
        OPENSSL_free(*buffer);
        OPENSSL_free(decoded_buffer);
        EVP_PKEY_CTX_free(ctx);

        return -5;
    }

    *out_len = output_length;

    OPENSSL_free(decoded_buffer);
    EVP_PKEY_CTX_free(ctx);

    return 0;
}
```

New main function instructions : 

```
int main(void){
        EVP_PKEY * pkey = EVP_RSA_gen(2048);

	char string[] = "HelloWorld";
	char * decrypted_buff;
        unsigned int decrypt_size = 0;
        int retval = 0;

        // ... Encryption part.
        // encrypted_buffer is a buffer containing the encrypted data.

        retval = RSADecrypt(pkey, encrypted_buffer, &decrypted_buffer, &decrypt_size, 2048 / 8);
	decrypted_buffer[decrypt_size] = '\0';
 
        if (retval != 0){
            free(decrypted_buff);

            printf("RSADecrypt failed with : %i\n", retval);
            return -1;
        }

        printf("> %s (%i characters were written)\n", decrypted_buffer, decrypt_size);

        free(decrypted_buff);

	return 0;
}
```

Note that i did not included the encryption part for simplicity matters, and since it ships base64 decoding in the process it may not be suitable for every implementations.

**Comment by Darnethal0z** [comment]
> I expect that the problem is that you initialized the context for an encrypt operation rather than a decrypt operation. Doing so sets the ctx->operation value to EVP_PKEY_OP_ENCRYPT. When you then call EVP_PKEY_decrypt, it compares the set operation to EVP_PKEY_OP_DECRYPT, and, when it doesn't match, returns -1.
> 
> The documentation for EVP_PKEY_decrypt: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_decrypt.html Indicates that a failure causes the function to return _either_ 0 or a negative value, but you only check for zero in the code above. As such you skip your error condition, and continue on as if things had worked properly.
> 
> I imagine that if you change the call to EVP_PKEY_encrypt_init to be EVP_PKEY_decrypt_init, things will work better.
> 
> Hope that helps!

Thank you for your answer, yes indeed i used the wrong init function. I used a different approach on my side as described in [my message](https://github.com/openssl/openssl/issues/22825#issuecomment-1826894720).
